ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,440 @@
+    }
+
+    // is Channel Format valid?
+    i2s_channel_fmt_t i2s_channelformat = args[ARG_channelformat].u_int;
+    if ((i2s_channelformat != I2S_CHANNEL_FMT_RIGHT_LEFT) &&
+        (i2s_channelformat != I2S_CHANNEL_FMT_ALL_RIGHT) &&
+        (i2s_channelformat != I2S_CHANNEL_FMT_ALL_LEFT) &&
+        (i2s_channelformat != I2S_CHANNEL_FMT_ONLY_RIGHT) &&
+        (i2s_channelformat != I2S_CHANNEL_FMT_ONLY_LEFT)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("Channel Format is not valid"));
+    }
+
+    // is Sample Rate valid?
+    // No validation done:  ESP-IDF API does not indicate a valid range for sample rate
+
+    // is DMA Buffer Count valid?
+    // ESP-IDF API code checks for buffer count in this range:  [2, 128]
+    int16_t i2s_dmacount = args[ARG_dmacount].u_int;
+    if ((i2s_dmacount < 2) || (i2s_dmacount > 128)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("DMA Buffer Count is not valid.  Allowed range is [2, 128]"));
+    }
+
+    // is DMA Buffer Length valid?
+    // ESP-IDF API code checks for buffer length in this range:  [8, 1024]
+    int16_t i2s_dmalen = args[ARG_dmalen].u_int;
+    if ((i2s_dmalen < 8) || (i2s_dmalen > 1024)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("DMA Buffer Length is not valid.  Allowed range is [8, 1024]"));
+    }
+
+    // is APLL Rate valid?
+    // No validation done:  ESP-IDF API does not indicate a valid range for APLL rate
+    
+    self->bck = bck;
+    self->ws = ws;
+    self->sdout = sdout;
+    self->sdin = sdin;
+    self->standard = args[ARG_standard].u_int;
+    self->mode = args[ARG_mode].u_int;
+    self->dataformat = args[ARG_dataformat].u_int;
+    self->channelformat = args[ARG_channelformat].u_int;
+    self->samplerate = args[ARG_samplerate].u_int;
+    self->dmacount = args[ARG_dmacount].u_int;
+    self->dmalen = args[ARG_dmalen].u_int;
+    self->apllrate = args[ARG_apllrate].u_int;
+
+    i2s_config_t i2s_config;
+    i2s_config.communication_format = self->standard;
+    i2s_config.mode = self->mode;
+    i2s_config.bits_per_sample = self->dataformat;
+    i2s_config.channel_format = self->channelformat;
+    i2s_config.sample_rate = self->samplerate;
+    i2s_config.intr_alloc_flags = ESP_INTR_FLAG_LOWMED; // allows simultaneous use of both I2S channels
+    i2s_config.dma_buf_count = self->dmacount;
+    i2s_config.dma_buf_len = self->dmalen;
+    if (self->apllrate != 0) {
+        i2s_config.use_apll = true;
+    } else {
+        i2s_config.use_apll = false;
+    }
+    i2s_config.fixed_mclk = self->apllrate;
+
+    // uninstall I2S driver when changes are being made to an active I2S peripheral
+    if (self->used) {
+        i2s_driver_uninstall(self->id);
+    }
+
+    esp_err_t ret = i2s_driver_install(self->id, &i2s_config, 0, NULL);
+    switch (ret) {
+        case ESP_OK:
+            break;
+        case ESP_ERR_INVALID_ARG:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install:  Parameter error"));
```

Comment:
Hey Mike,

FYI

I encountered compilation errors that I traced back to the double-spaces in your error text messages (in this line for example the double space between "install:" and "Parameter").

Specifically: I'm running my own branch from v1.13 and hand-edited the one-line changes to modmachine.h, modmachine.c and the Makefile, and downloaded machine_i2s.c into the code base.

Launching the compilation process gave me the following error:
...
GEN build-TINYPICO/genhdr/compressed.collected
QSTR not updated
Compressed data updated
GEN build-TINYPICO/genhdr/compressed.data.h
Traceback (most recent call last):
  File "../../py/makecompresseddata.py", line 205, in <module>
    main(sys.argv[1], word_compression)
  File "../../py/makecompresseddata.py", line 165, in main
    compressed_data = fn(error_strings)
  File "../../py/makecompresseddata.py", line 78, in word_compression
    return "".join(w[:-1] + "\\{:03o}".format(0b10000000 | ord(w[-1])) for w in index)
  File "../../py/makecompresseddata.py", line 78, in <genexpr>
    return "".join(w[:-1] + "\\{:03o}".format(0b10000000 | ord(w[-1])) for w in index)
IndexError: string index out of range
...

I traced this down to the index array variable inside makecompresseddata.py whose last element was empty quotes '' instead of a keyword. This empty quotes value itself comes from the top128 array a few lines earlier where there are 12 instances of this empty quotes values counted in the collection. The 12 instances relate to the 12 instances of double-spacing in the error messages of machine_i2s.c. Replacing those 12 instances with single spaces resolved the compilation issue.

top128 prints out as:
[('invalid', 30), ("can't", 35), ('must', 30), ("'%q'", 27), ('not', 32), ('object', 18), ('argument', 14), ('arguments', 11), ('function', 12), ('format', 15), ('be', 30), ('error', 15), ('convert', 11), ('for', 20), ('to', 25), ('Wifi', 15), ("'%s'", 14), ('keyword', 9), ('required', 8), ('of', 20), ('I2S', 15), ('string', 9), ('is', 19), ('specifier', 6), ('supported', 6), ('type', 11), ('index', 9), ('length', 7), ('range', 8), ('Invalid', 6), ('with', 9), ('%d', 14), ('Undocumented', 4), ('in', 14), ('arg', 10), ('empty', 7), ('implemented', 4), ("isn't", 7), ('multiple', 5), ('a', 18), ('assignment', 4), ('identifier', 4), ('or', 12), ('positional', 4), ('already', 5), ("doesn't", 5), ('Parameter', 4), ('UUID', 7), ('attribute', 4), ('expecting', 4), ('generator', 4), ('out', 8), ('syntax', 5), ('too', 8), ('expected', 4), ('from', 6), ('SPI', 7), ('allowed', 4), ('an', 9), ('and', 7), ('complex', 4), ('field', 5), ('has', 7), ('integer', 4), ('missing', 4), ('needs', 5), ('no', 9), ('outside', 4), ('set', 7), ('tuple', 5), ('unknown', 4), ('unsupported', 3), ('value', 5), ('%s', 8), ('Failed', 4), ('buffer', 4), ('memory', 4), ('number', 4), ('int', 6), ('name', 5), ('operation', 3), ('pin', 6), ('specified', 3), ('Error', 4), ('Internal', 3), ('found', 4), ('install:', 3), ('nonlocal', 3), ('param', 4), ('sequence', 3), ('valid', 4), ('Not', 5), ('bad', 5), ('between', 3), ('but', 5), ('indices', 3), ('struct:', 3), ('%q', 6), ('args', 4), ('characteristic', 2), ('dict', 4), ('list', 4), ('long', 4), ('mode', 4), ('on', 6), ('only', 4), ('zero', 4), ('Cannot', 3), ('TCP/IP', 3), ('before', 3), ('driver', 3), ('import', 3), ('method', 3), ('specification', 2), ('values', 3), ('incompatible', 2), ('issubclass()', 2), ('after', 3), ('bytes', 3), ('class', 3), ('given', 3), ('key', 4), ('local', 3), ('non-keyword', 2), ('store', 3), ('use', 4), ('wrong', 3), ('', 12)]

and index prints out as:
['invalid', "can't", 'must', "'%q'", 'not', 'object', 'argument', 'arguments', 'function', 'format', 'be', 'error', 'convert', 'for', 'to', 'Wifi', "'%s'", 'keyword', 'required', 'of', 'I2S', 'string', 'is', 'specifier', 'supported', 'type', 'index', 'length', 'range', 'Invalid', 'with', '%d', 'Undocumented', 'in', 'arg', 'empty', 'implemented', "isn't", 'multiple', 'a', 'assignment', 'identifier', 'or', 'positional', 'already', "doesn't", 'Parameter', 'UUID', 'attribute', 'expecting', 'generator', 'out', 'syntax', 'too', 'expected', 'from', 'SPI', 'allowed', 'an', 'and', 'complex', 'field', 'has', 'integer', 'missing', 'needs', 'no', 'outside', 'set', 'tuple', 'unknown', 'unsupported', 'value', '%s', 'Failed', 'buffer', 'memory', 'number', 'int', 'name', 'operation', 'pin', 'specified', 'Error', 'Internal', 'found', 'install:', 'nonlocal', 'param', 'sequence', 'valid', 'Not', 'bad', 'between', 'but', 'indices', 'struct:', '%q', 'args', 'characteristic', 'dict', 'list', 'long', 'mode', 'on', 'only', 'zero', 'Cannot', 'TCP/IP', 'before', 'driver', 'import', 'method', 'specification', 'values', 'incompatible', 'issubclass()', 'after', 'bytes', 'class', 'given', 'key', 'local', 'non-keyword', 'store', 'use', 'wrong', '']

I'm no expert, and am open to the idea that I have some other local (self-induced!) issue causing this error - but I'm surprised this compilation error has not happened with other users since the 12 instances should appear in most compilation runs and therefore manifest the error. This is probably more a case of some latent fragility within makecompresseddata.py (a split() on space gone wrong?) that only manifests when enough instances of this empty string make it into the top 128 words; but it seems easy enough to remove the double-spacing in your error messages to prevent future headaches anyway?

Hope this helps! Thanks for your I2S code! - I look forward to giving it a spin ...

Reply:
Thanks for taking the time to report this problem.  I changed all instances of double-spaces to single-spaces, then rebased to the latest commit.  Hopefully this change fixes the build problem.